Make wxMSW date/time picker controls locale aware

utelle opened the pull request:
- Add helper function to retrieve native format specs from wxUILocale
- Add function declaration of the helper function in msw/private/uilocale.h
- Adjust code for date/time picker to set native format specification

utelle committed:
Make wxMSW date/time picker controls locale aware

- Add helper function to retrieve native format specs from wxUILocale
- Add function declaration of the helper function in msw/private/uilocale.h
- Adjust code for date/time picker to set native format specification

utelle commented:
On the [wxWidgets forum](https://forums.wxwidgets.org/viewtopic.php?f=23&p=221882&sid=813335fad176f8a66d5c300d894a9048#p221882) someone asked recently about locale aware date formatting in the date picker control.

Will it be considered to merge this PR? AFAICT the failing checks are not related to this PR but to other issues.

vadz commented:
Sorry, I never found time to get back to this and finally understand what can, and should, we do here. I'll try to do it before the end of the year...

vadz commented:
(same apology as for the other PR)

Do I understand correctly that this implements using the locale set with `wxUILocale::UseLocaleName()` for the calendar control? Does it change the default behaviour, i.e. what happens if neither this function nor `wxUILocale::UseDefault()` was called?

utelle committed:
Fix locale name if C locale is current locale

Helper function `wxGetMSWDateTimeFormat` will use the locale name `en-US` if the current locale name is `C`(the standard C locale) to avoid that Windows API function `GetLocaleInfoEx` will return an error.

utelle commented:
> Do I understand correctly that this implements using the locale set with `wxUILocale::UseLocaleName()` for the calendar control?

Well, the calendar control where you see the month name and the abbreviated weekday names will stay in the default locale of the system (German on my system). This can't be changed unfortunately. However, date controls where the date is displayed as a string will show the localized string. For example `16.10.2024` (German) or `16/10/2024` (French) or `10/16/2024` (US English).

> Does it change the default behaviour, i.e. what happens if neither this function nor `wxUILocale::UseDefault()` was called?

When I started the PR I had not tested that case. But I did now. If neither `wxUILocale::UseLocaleName()` nor `wxUILocale::UseDefault()` was called, the `wxUILocale::GetCurrent()` will return the standard C locale. The Windows API function `GetLocaleInfoEx()` will then be called with the name of the standard C locale. However, the locale name `C` is unknown and the function will return an error. This should not happen, of course. Therefore I applied a small modification to use the locale name `en-US` in that case.

utelle commented:
> Thanks for the change but there are new warnings now, could you please check them? TIA!

There are compile errors reported in the CI check _MSW cross-builds / wxMSW 32 bits_ and _MSW cross-builds / wxMSW 32 bits_. However, the problem is that I don't see how those errors are related to the changes I made. Here are the messages from the CI run log:

```
In file included from ../include/wx/msw/private/uilocale.h:13,
                 from ../include/wx/msw/timectrl.h:14,
                 from ../include/wx/timectrl.h:98,
                 from ./allheaders.h:346,
                 from ./allheaders.cpp:371:
../include/wx/msw/private.h: In function ‘COLORREF wxColourToRGB(const wxColour&)’:
../include/wx/msw/private.h:225:12: error: conversion to ‘long unsigned int’ from ‘int’ may change the sign of the result [-Werror=arith-conversion]
  225 |     return RGB(c.Red(), c.Green(), c.Blue());
      |            ^~~
../include/wx/msw/private.h: In function ‘COLORREF wxColourToPalRGB(const wxColour&)’:
../include/wx/msw/private.h:230:12: error: conversion to ‘long unsigned int’ from ‘int’ may change the sign of the result [-Werror=arith-conversion]
  230 |     return PALETTERGB(c.Red(), c.Green(), c.Blue());
      |            ^~~~~~~~~~
../include/wx/msw/private.h: At global scope:
../include/wx/msw/private.h:1020:81: error: zero as null pointer constant [-Werror=zero-as-null-pointer-constant]
 1020 | WXDLLIMPEXP_BASE wxString wxMSWFormatMessage(DWORD nErrCode, HMODULE hModule = 0);
      |                                                                                 ^
In file included from ../include/wx/object.h:23,
                 from ../include/wx/event.h:15,
                 from ../include/wx/evtloop.h:13,
                 from ./testprec.h:5,
                 from ./allheaders.cpp:369:
../include/wx/string.h: In instantiation of ‘int wxString::Printf(const wxFormatString&, Targs ...) [with Targs = {const char*, int, const wchar_t*, long int, wxString}]’:
../include/wx/string.h:2444:15:   required from ‘static wxString wxString::Format(const wxFormatString&, Targs ...) [with Targs = {const char*, int, const wchar_t*, long int, wxString}]’
../include/wx/log.h:999:37:   required from ‘void wxLogger::Log(const FormatString&, Targs ...) [with Targs = {const char*, int, const wchar_t*, long int, wxString}]’
../include/wx/msw/private.h:370:9:   required from here
../include/wx/string.h:2422:54: error: ‘value’ is not a member of ‘wxFormatStringSpecifier<const char*>’
 2422 |     format.Validate({wxFormatStringSpecifier<Targs>::value...});
      |                                                      ^~~~~
../include/wx/string.h:2422:20: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘const std::vector<int>&’
 2422 |     format.Validate({wxFormatStringSpecifier<Targs>::value...});
      |     ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../include/wx/string.h:37:
../include/wx/strvararg.h:212:43: note:   initializing argument 1 of ‘void wxFormatString::Validate(const std::vector<int>&) const’
  212 |     void Validate(const std::vector<int>& argTypes) const;
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
../include/wx/string.h:2432:85: error: ‘const char* wxArgNormalizer<const char*>::get() const’ is private within this context
 2432 |       return DoPrintfWchar(format, wxArgNormalizerWchar<Targs>{args, nullptr, 0}.get()...);
      |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
../include/wx/strvararg.h:689:17: note: declared private here
  689 |     const char *get() const;
      |                 ^~~
cc1plus: all warnings being treated as errors
```

Then _ci/circleci build-fedora_ failed. The step _Build tests_ failed. Again I have no idea how this is related to my changes. There are tons of error messages. Find below the first few of them:

```
In file included from ../include/wx/evtloop.h:14,
                 from ./testprec.h:5,
                 from ./allheaders.cpp:369:
../include/wx/event.h: In copy constructor 'wxSetCursorEvent::wxSetCursorEvent(const wxSetCursorEvent&)':
../include/wx/event.h:1944:11: error: implicitly-declared 'wxCursor::wxCursor(const wxCursor&)' is deprecated [-Werror=deprecated-copy-dtor]
 1944 |           m_cursor(event.m_cursor)
      |           ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../include/wx/cursor.h:51,
                 from ../include/wx/event.h:22:
../include/wx/gtk/cursor.h:37:13: note: because 'wxCursor' has user-provided 'virtual wxCursor::~wxCursor()'
   37 |     virtual ~wxCursor();
      |             ^
../include/wx/event.h: In member function 'void wxSetCursorEvent::SetCursor(const wxCursor&)':
../include/wx/event.h:1950:57: error: implicitly-declared 'wxCursor& wxCursor::operator=(const wxCursor&)' is deprecated [-Werror=deprecated-copy-dtor]
 1950 |     void SetCursor(const wxCursor& cursor) { m_cursor = cursor; }
      |                                                         ^~~~~~
../include/wx/gtk/cursor.h:37:13: note: because 'wxCursor' has user-provided 'virtual wxCursor::~wxCursor()'
   37 |     virtual ~wxCursor();
      |             ^
```

vadz commented:
Sorry, I hadn't checked the warnings details, but after doing it I realize that this is indeed rather mysterious, I wonder if it's due to some fixes done in master since this PR was created but not present in it. I've created #24908 with your changes to check if the CI builds pass there and I'll merge that one if they do and will look at this more carefully if they don't.

utelle commented:
> Oh, wait, I know what's wrong here: the errors are due to including `include/wx/msw/private.h` from a public header (`include/wx/timectrl.h`).

Ouch. Sorry, my bad. I should have paid attention to that.

> It's actually lucky that we got them because otherwise this could have gone unnoticed,

Yes, indeed. Thanks for detecting the reason for the compile time errors.

> but we're not supposed to include private headers from public ones, as this makes them de facto public.
> 
> The simplest fix is to move the code using the private function to the source file, I'll do this in the other PR and will close this one to keep only one of them open.

Ok. Thank you.
